### ISSUE TITLE
Generalize sublist relation to subset relation

### ISSUE THREAD

**Issue by HuStmpHrrr** [agda/agda-stdlib#292]
I've been thinking to use sublist relation https://github.com/agda/agda-stdlib/blob/master/src/Data/List/Relation/Sublist/Setoid.agda `_⊆_` for something. However, I soon realized that it's defined to be sublist, instead of subset, where the latter is in fact what I want.

By checking the definition, there really isn't a reason, at the point of defining `_⊆_`, to conclude it's a sublist or subset. In fact, it's a sublist relation, only after we specify what equivalence relation we are talking about. Currently, it's only clear in https://github.com/agda/agda-stdlib/blob/master/src/Data/List/Relation/Sublist/Setoid/Properties.agda , which defines a preorder relation using pointwise equivalence, which is more restrictive. 

Therefore, I propose to make this more generalized, to make the relation can be used in both ways.

If you also agree it's a good move, I don't mind to contribute to this issue.

**Comment by HuStmpHrrr** [comment]
@MatthewDaggitt  
consider sublist and subset relations defined over `List`. it's easy to realize that sublist subsumes subset, meaning that two lists can form a subset relation but not sublist relation.

Consider following equivalence relation,

```
l1 ~~ l2 = l1 ⊆ l2 x l2 ⊆ l1
```

we can prove it's a equivalence relation by routine check. If we define a preorder over it with `_⊆_`, we obtain a weaker relation. This is useful, and align better with membership relation, as if one reasons membership/containment, he should not require the preorder relation over sublists but rather subsets.

**Comment by MatthewDaggitt** [comment]
Ah! I now see what you mean.

The current `Data.List.Relation.Equality` hierarchy is nowhere near deep enough *sigh*

Okay so tentatively can I propose the following module structure inside `Data.List.Relation.Equality`?
- Pointwise
  * Setoid (exists)
  * DecSetoid (exists)
  * Propositional (exists)
  * DecPropositional (exists)
- Set
  * Relational (exists, the set part of the current `BagAndSetEquality`)
  * MEANINGFUL_NAME (new, @HuStmpHrrr's definition above)
     - Setoid
     - DecSetoid
     - Propositional
     - DecPropositional
- Bag
  * Relational (exists, the bag part of the current `BagAndSetEquality` file)

Thoughts? I'm not so happy with the `Relational` name for @nad's work.... Maybe he can suggest a better one?

**Comment by HuStmpHrrr** [comment]
actually this can be more general, all preorder induces an equivalence relation and thereafter comes a partial order relation.

https://github.com/agda/agda-stdlib/blob/f0d2286097369fd3eb4863f37d63b9f1a9b4bdc2/src/Relation/Binary/Properties/Preorder.agda

Another option is to reinforce this module and then instantiate in `Data.List`.

**Comment by MatthewDaggitt** [comment]
I think this is closed this with bb8b28eda66ea1d9c14a6cc30b76462c8ac41b12.

@HuStmpHrrr feel free open a new issue for the set interface if you'd like.

**Comment by nad** [comment]
I took a look at [`Data.List.Relation.Sublist.Extensional.Setoid`](https://github.com/agda/agda-stdlib/blob/8b7490b1c57702f86ef2936a0e52c22740b2f05d/src/Data/List/Relation/Sublist/Extensional/Setoid.agda). In what sense is the relation given there a sublist relation?

**Comment by MatthewDaggitt** [comment]
Hmm you're right. It's more a subset relation than a sublist relation. This hierarchy is tying me up knots.

**Comment by nad** [comment]
When I first defined (7ac3a369390f2711723ac3ba4d0038292a9b3636) the relation that is now in `Data.List.Relation.Sublist.Extensional.Setoid` I documented it in the following way:

https://github.com/agda/agda-stdlib/blob/7ac3a369390f2711723ac3ba4d0038292a9b3636/Data/List/Any.agda#L34-L42

**Comment by HuStmpHrrr** [comment]
sublist relation sounds like OPE defined by @gallais in `Inductive.agda`, doesn't it? I perceive sublist to be a specialized relation of subset which additionally requires order preserving. this restriction is only added as pointwise equivalence is introduced inside of the inner module. this is the reason why i opened this issue.

should we revise how we want to organize subset and sublist relations from module level? i would say let's start with subset then we gradually specialize more as we want to have more restrictive relations.

**Comment by MatthewDaggitt** [comment]
Closed with a14f9f1
